docs/decisions/infra/0005-database-migration-architecture.md

```diff
@@ -0,0 +1,96 @@
+# Database Migration Infrastructure and Deployment
+
+* Status: proposed
+* Deciders: [list everyone involved in the decision] <!-- optional -->
+* Date: [YYYY-MM-DD when the decision was last updated] <!-- optional -->
+
+## Context and Problem Statement
+
+What is the most optimal setup for database migrations infrastructure and deployment?
+This will break down the different options for how the migration is ran, but not the
+tools or languages the migration will be ran with, that will be dependent on the framework
+the application is using.
+
+Questions that need to be addressed:
+ 1. How will the method get the latest migration code to run?
+ 2. What infrastructure is required to use this method?
+ 3. How is the migration deployment re-ran in case of errors?
+
+## Decision Drivers <!-- optional -->
+
+* [driver 1, e.g., a force, facing concern, …]
+* [driver 2, e.g., a force, facing concern, …]
+* … <!-- numbers of drivers can vary -->
+
+## Considered Options
+
+* Ran via a Lambda Function
+* Ran via an ECS Fargate Task
+
+## Decision Outcome
+
+Chosen option: "[option 1]", because [justification. e.g., only option, which meets k.o. criterion decision driver | which resolves force force | … | comes out best (see below)].
+
+### Positive Consequences <!-- optional -->
+
+* [e.g., improvement of quality attribute satisfaction, follow-up decisions required, …]
+* …
+
+### Negative Consequences <!-- optional -->
+
+* [e.g., compromising quality attribute, follow-up decisions required, …]
+* …
+
+## Pros and Cons of the Options <!-- optional -->
+
+### Ran via a Lambda Function
+
+[example | description | pointer to more information | …] <!-- optional -->
+
+#### Pros
+
+
+
+#### Cons
+
+
+
+### Ran via an ECS Fargate Task
+
+In this method, the migration is ran by an ECS task that spins up in a cluster.
+This task will be destroyed once the migration is completed, so it will not need to be
+a long running service.
+
+This method gets the latest configuration by creating a Dockerized image from the
+workflow and pushing it to AWS ECR. This image has a function that will be called
+from the ECS task's configuration JSON.
+
+#### Infrastructure Required
+
+ - ECS cluster to run the task in, which can be the same cluster as the application
+ - Log group for storing logs
+ - IAM role for the task to use
+ - (Optional) IAM policy with permission to connection if using IAM authentication on the database
+ - (Optional) ECR to store the image. This is dependent on the application used
```

Comment:
how come the ECR repository is optional? how would it work if we didn't have an ECR repository?

Reply:
It's optional because of the app image you might use. For the Flask app we're using, we are using the same image for the app as the migrations, because we are running the `db-migrate-*` make commands that we do locally. There's no point building and saving a different image in that case because it's the same code effectively. We just use a command override in the config for the migration so the task knows to run the migration command vs the normal command that runs the API

Reply:
oh yeah that makes sense to me and what i would have imagined. i'd actually push for making that approach the requirement – i.e. we should use the same image as the application, and that image should have a command `db-migrate` that can be called from the task runner to override the default command that is used for the web service.